src/python/pants/backend/go/module.py

```diff
@@ -174,6 +179,44 @@ async def resolve_go_module(
     )
 
 
+@dataclass(frozen=True)
+class FindOwningGoModuleRequest:
+    address: Address
+
+
+@dataclass(frozen=True)
+class ResolvedOwningGoModule:
+    module_address: Optional[Address]
+
+
+@rule
+async def find_nearest_go_module(request: FindOwningGoModuleRequest) -> ResolvedOwningGoModule:
+    # Obtain unexpanded targets and ensure file targets are filtered out. Unlike Python, file targets do not
+    # make sense semantically for Go source since Go builds entire packages at a time. The filtering is
+    # accomplished by requesting `UnexpandedTargets` and also filtering on `is_file_target`.
+    spec_path = request.address.spec_path
```

Comment:
Since only the `spec_path` of the `Address` is used, `FindOwningGoModuleRequest` should only contain the `spec_path`. That will improve memoization by causing all lookups within each directory to be memoized once (currently you'd get a distinct copy memoized per `Address` within a directory).

Reply:
Done.